### PR TITLE
Introduce `ThinContext` trait for static `Context`s and zero sized types

### PR DESCRIPTION
**Pull request by mkatychev**
## 🌟 What is the purpose of this PR?

A lot of of the value in reports is the separation of error type from metadata (contexts & attachments).
In my daily use of error-stack, I found the greatest value when the `Context` was a zero sized type, allowing things like seamless conversions of [`Option<T>` into `Report<NotFound>`](https://github.com/knox-networks/bigerror/blob/1e9c47758e4c311c4a79866b0f57f58991861857/bigerror/src/lib.rs#L465-L476) types if `NotFound` was a zero sized type.

This PR upstreams the `ThinContext` trait to allow behaviour that is only possible when a `Context` is a static value at compile time:
https://github.com/hashintel/hash/blob/5aeb3bfa66a56c9c3f7d5ee7468a4fcdd7275b7a/libs/error-stack/src/context.rs#L134-L145

This allows a report to get a value implicitly when calling `Report::change_context` greatly increasing the ergonomics of propagating errors as well as incrementing just the location of when an error could propagate multiple times:
`err.change_context(MyError)` becomes `err.into_ctx()`

## 🔗 Related links

Fork of error-stack with `ThinContext` implemented: https://github.com/knox-networks/bigerror

- ...

## 🚫 Blocked by

This adds a `#[derive(ThinContext)]` macro to the error-stack macro crate, it will need to be published for error-stack to feature this on crates.io

## 🔍 What does this change?

* Feature addtion to `error-stack` to scope behaviour for a subset of contexts

### 🚢 Has this modified a publishable library?

This PR:

- [x] modifies a **Cargo**-publishable library, but **it is not yet ready to publish**

### 📜 Does this require a change to the docs?

yes 

The changes in this PR:

- [x] require changes to docs which are **not** made in this PR

### 🕸️ Does this require a change to the Turbo Graph?

The changes in this PR:

- [x] I am unsure / need advice

## 🐾 Next steps

The test coverage and documentation are lacking in this branch currently. This PR is intended to test the waters for this kind of feature.

## 🛡 What tests cover this?

* `error-stack/tests/thin_ctx.rs`

-

## ❓ How to test this?

1. run `cargo test` in error-stack

## 📹 Demo
[The test code](https://github.com/knox-networks/bigerror/blob/1e9c47758e4c311c4a79866b0f57f58991861857/bigerror/src/lib.rs#L675-L685):

```rust
#[test]
fn result_into_ctx() {
    fn output() -> Result<usize, Report<MyError>> {
        "NaN"
        .parse::<usize>()
        .map_err(ConversionError::from::<&str, usize>)
        .into_ctx()
        .attach_printable(
            "Report::into_ctx<MyError> will be called twice without creating an extra Frame!",
        )
    }

    assert_err!(output().into_ctx::<MyError>());
}
```

image

### CODE REVIEW

File: libs/error-stack/macros/Cargo.toml [hunk 1]
```diff
@@ -1,25 +1,21 @@
 [package]
-name = "error-stack-macros"
-version = "0.0.0-reserved"
-authors = { workspace = true }
+name = "error-stack-derive"
```
Comment (TimDiekmann) [inline libs/error-stack/macros/Cargo.toml:2]: Could you elaborate why you rename a crate owned by us? The crate is [published on `crates.io`](https://crates.io/crates/error-stack-macros) already. You did not even mention this in the PR description!

**Review by TimDiekmann — CHANGES_REQUESTED** [review]
Hi!

I'm not entirely sure if I get the reason behind this implementation. Is the goal to simplify the generation of `Report`s? The example you gave is to create `Report<NotFound>` from an `Option<T>`. This would result in quite a few `OptionExt`-like traits (which we explicitly don't want to implement in `error-stack`: https://github.com/hashintel/hash/issues/4777#issuecomment-2275377679), which makes usage quite hard to follow. Also, I don't see why `into_ctx` needs to be called twice.

File: libs/error-stack/macros/Cargo.toml (hunk 1 above)
Comment (mkatychev) [reply libs/error-stack/macros/Cargo.toml:2]: That's an oversight, I'll revert

**Comment by mkatychev** [comment]
> Hi!
> 
> I'm not entirely sure if I get the reason behind this implementation. Is the goal to simplify the generation of `Report`s? The example you gave is to create `Report<NotFound>` from an `Option<T>`. This would result in quite a few `OptionExt`-like traits (which we explicitly don't want to implement in `error-stack`: [#4777 (comment)](https://github.com/hashintel/hash/issues/4777#issuecomment-2275377679)), which makes usage quite hard to follow. Also, I don't see why `into_ctx` needs to be called twice.

Perhaps the `NotFound` was a poor example. I don't think introducing error types into error-stack is a good idea for the reasons you mentioned (the `OptionExt` trait may need some rethinking or not even used at all, they are valid points).

The intent of this PR was to allow ZSTs and const types to have a smaller overhead and to allow adding attachments rather than growing the context depth if you have a lot of functions that produce the same `Report<MyZstError>`s through nested function calls.

Errors whose value is known at compile time have no need to increment context depth so long as we are propagating the same type  so something like `Report<MyZstErr>` -> `Report<MyZstErr>` -> `Report<MyZstErr>` would give a clearer picture with this kind of output:

```
MyZstErr
- location.rs 10:10
- location_2.rs 100:10
- location.rs 20:10
```

Over the current approach:
```
MyZstErr
- location.rs 10:10
  MyZstErr
  - location_2.rs 100:10
  MyZstErr
  - location.rs 20:10
```

**Comment by mkatychev** [comment]
I've updated the PR to reduce scope, address feedback, and elaborate the test to clarify the use case for a `ThinContext`.

**Comment by indietyp** [comment]
Thank you for your PR but I share the sentiment of @TimDiekmann, I don't think it really belongs into error-stack (and looking at the implementation it actually doesn't even require access to error-stack internals). I would even go as far as saying that something like this goes against the philosophy of error-stack. Let me explain:

In error-stack we made sure that everything is explicit and deliberate - every single change that you do to a report - such as changing the context are deliberate and explicit - this removes that explicitness moving from a: this always changes the context to: "this sometimes changes the context under certain circumstances" it also goes against the philosophy in the sense that every context change is deliberate as it delineates a sort of "border" - the types from this part of the code-base return this error.

To quote the documentation:

> The current context should encapsulate how the current code interprets the error. As the error propagates, it will cross boundaries where new information is available, and the previous level of detail is no longer applicable. These boundaries will often occur when crossing between major modules, or when execution crosses between crates.

Your implementation fundamentally changes this allowing for a `change_context(A).change_context(A)`, this is fine if we cross a boundary or it is intentional, but your code makes this boundary conditional (in my personal opinion it's also a code-smell, if you have an error of `Report<C>` and you want to propagate it why call `change_context` at all if the "border" you use doesn't change).

This is it from a philosophical standpoint. From a code standpoint there are two major questions that arise for me:

A) why is a derive macro needed? Don't we already have `thiserror` or `derive-more` (I would kinda understand this if the point was to e.g. take advantage of things that aren't implemented, such as proper support of the `provide` API, but that's another discussion)
B) your code doesn't do what it says.
C) why `ctx` if everywhere else we deliberately use `context`

You intend that `ThinContext` is used only for ZSTs, but that isn't the case, nothing in the trait prohibits (I also don't know _if_ there is even a way to enforce that a trait be only implemented on ZSTs) but instead _any_ type that is constructable in a const context, using that assumption also automatically invalidates the trait implementation. What happens if my current context is `JsonError::InvalidChar` and I want to change it to `JsonError::UnexpectedLBrace`, `JsonError` is constructable in a const context and implements (in this example) `ThinContext`, if I now use the conditional `change_context` my change to `JsonError::UnexpectedLBrace` would simply be ignored and that would be super frustrating to debug, is hidden unintended behaviour.

**Comment by mkatychev** [comment]
> Your implementation fundamentally changes this allowing for a `change_context(A).change_context(A)`, this is fine if we cross a boundary or it is intentional, but your code makes this boundary conditional (in my personal opinion it's also a code-smell, if you have an error of `Report<C>` and you want to propagate it why call `change_context` at all if the "border" you use doesn't change).

While the context does not change the location that it is propagated through does, if a ZST context is propagated through may functions there is often only one `Location` recorded making the `?` operator be the non canonical way to use error-stack. This PR doesn't fix that problem (perhaps this will be easier with try_v2 🤞)

> B) your code doesn't do what it says.

I don't know how to reply to this statement, could you add some context?

> Instead `ThinContext` allows _any_ type that is constructable in a const context

that was the intent of the derive throwing an err on an enum, I don't have a solution to that other than not allowing than to not have enums imlement `ThinContext` in the same way a user would not implement `Default` on an enum.

I appreciate the warm discussion and it makes sense for this code to not live in error-stack, the gap of propagating a context whilst only attaching new location still exists, I realize that one does not _change_ the context but there doesn't seem (in stable) a way to only add a location to a report before propagation, perhaps a `Report::add_location` public function could be used?

Thanks for making a great error library!

**Comment by mkatychev** [comment]
Thanks for the discussion @indietyp @TimDiekmann, appreciate the courteous feedback!
